README.md
utf8_validator.lua
=================

This Lua function can be used to validate UTF-8 strings.

Usage:

    local utf8_validator = require 'utf8_validator'

    local is_valid  = utf_validator.validate('This is valid')

    -- you can also use the shortcut syntax:
    local is_valid2 = utf_validator('This is also valid')

If a string is not valid utf8, then `validate` will also return the position of the first byte that makes it non-valid:

    local is_valid, error_position = utf_validator(a_binary_blob)
    if not is_valid then print('non-utf8 sequence detected at position ' .. tostring(error_position))

For valid utf8 strings, `error_position` is always `nil`.


Installation
============

Just copy the `utf8_validator.lua` file wherever you need it and `require` it.

License
=======

This library is MIT-licensed.

Specs
=====

This library uses [busted](http://olivinelabs.com/busted) for its specs. In order to run them, install `busted` and then:

    cd path/where/the/spec/folder/is
    busted






CHANGELOG.md
# 0.0.2

* Fixed erroneous limit for 2-byte characters (#2)


spec/utf_validator_spec.lua
local utf = require 'utf_validator'

describe('utf_validator', function()
  it('exists', function()
    assert.is_truthy(utf)
  end)

  describe('utf_validator.getEncoding', function()

    it('detects ascii', function()
      local e,n = utf.getEncoding('Hello word')
      assert.equal(e, 'ascii')
      assert.is_nil(n)
    end)

    it('#focus detects utf8', function()
      local e,n = utf.getEncoding('¢€𤭢')
      assert.equal(e, 'utf-8')
      assert.is_nil(n)
    end)

    it('returns utf8 if ascii is merged with utf8', function()
      local e,n = utf.getEncoding('I will pay in €. Thank you.')
      assert.equal(e, 'utf-8')
      assert.is_nil(n)
    end)
  end)
end)


spec/utf8_validator_spec.lua
local utf8 = require 'utf8_validator'

describe('utf8_validator', function()
  it('exists', function()
    assert.is_truthy(utf8)
  end)

  local binary_str = assert(io.open('spec/banana.gif', "rb")):read("*all")

  local scenarios = {
    { desc = 'plain ascii', str = 'Hello word',        valid = true },
    { desc = 'utf8 stuff',  str = '¢€𤭢',              valid = true },
    { desc = 'mixed stuff', str = 'Pay in €. Thanks.', valid = true },
    { desc = 'binary',      str = binary_str,          valid = false, where=11 }
  }

  describe('utf8_validator.validate', function()
    for _,scenario in ipairs(scenarios) do
      it(scenario.desc, function()
        local is_valid, where = utf8.validate(scenario.str)
        assert.equal(scenario.valid, is_valid)
        assert.equal(scenario.where, where)
      end)
    end
  end)

  describe('utf8_validator.validate shortcut version', function()
    for _,scenario in ipairs(scenarios) do
      it(scenario.desc, function()
        local is_valid, where = utf8(scenario.str)
        assert.equal(scenario.valid, is_valid)
        assert.equal(scenario.where, where)
      end)
    end
  end)


end)


utf8_validator.lua
local utf8_validator = {
  __VERSION     = '0.0.1',
  __DESCRIPTION = 'Library for easily validating UTF-8 strings in pure Lua',
  __URL         = 'https://github.com/kikito/utf8_validator.lua',
  __LICENSE     = [[
    MIT LICENSE

    Copyright (c) 2013 Enrique García Cota

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  ]]
}

-- Taken from table 3-7 in www.unicode.org/versions/Unicode6.2.0/UnicodeStandard-6.2.pdf
local VALID_RANGES = {
  { {1, 127},                                       size=1 },
  { {194, 223}, {123, 191},                         size=2 },
  { {224, 224}, {160, 191}, {128, 191},             size=3 },
  { {225, 236}, {128, 191}, {128, 191},             size=3 },
  { {237, 237}, {128, 159}, {128, 191},             size=3 },
  { {238, 239}, {128, 191}, {128, 191},             size=3 },
  { {240, 240}, {144, 191}, {128, 191}, {128, 191}, size=4 },
  { {241, 243}, {128, 191}, {128, 191}, {128, 191}, size=4 },
  { {244, 244}, {128, 143}, {128, 191}, {128, 191}, size=4 }
}

local function matchesRow(row, str, from, to)
  for i=from, to do
    local byte = str:byte(i,i)
    local range = row[i-from+1]
    if range[2] < byte or byte < range[1] then
      return false
    end
  end
  return true
end

local function validateChar(str, from, bytes_size)
  for _,row in ipairs(VALID_RANGES) do
    local to = from + row.size - 1
    if to <= bytes_size then
      if matchesRow(row, str, from, to) then
        return true, row.size
      end
    end
  end
  return false, 1
end

function utf8_validator.validate(str)
  local bytes_size = #str
  local from = 1

  repeat
    local is_valid, size = validateChar(str, from, bytes_size)
    if not is_valid then return false, from end
    from = from + size
  until from > bytes_size

  return true
end

setmetatable(utf8_validator, {__call = function(_, ...) return utf8_validator.validate(...) end})

return utf8_validator


utf_validator.lua
local utf_validator = {
  __VERSION     = '0.0.1',
  __DESCRIPTION = 'Library for easily validating UTF strings in pure Lua',
  __URL         = 'https://github.com/kikito/utf_valiator.lua',
  __LICENSE     = [[
    MIT LICENSE

    Copyright (c) 2013 Enrique García Cota

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  ]]
}

local ENCODINGS = {
  { name = 'ascii',
    size = 1,
    ranges = {
      { {1, 127} }
    }
  },
  { name = 'utf-8',
    size = 2,
    ranges = {
      { {194, 223}, {123, 191} },
    }
  },
  { name = 'utf-8',
    size = 3,
    ranges = {
      { {224, 224}, {160, 191}, {128, 191} },
      { {225, 236}, {128, 191}, {128, 191} },
      { {237, 237}, {128, 159}, {128, 191} },
      { {238, 239}, {128, 191}, {128, 191} }
    }
  },
  { name = 'utf-8',
    size = 4,
    ranges = {
      { {240, 240}, {144, 191}, {128, 191}, {128, 191} },
      { {241, 243}, {128, 191}, {128, 191}, {128, 191} },
      { {244, 244}, {128, 143}, {128, 191}, {128, 191} }
    }
  }


}

local function matchRanges(ranges, bytes, from, to)
  local matching = {}
  for i=from, to do
    matching[#matching + 1] = bytes[i]
  end

  for i=from, to do
    local byte, range = bytes[i], ranges[i-from+1]
    if not (range[1] <= byte and byte <= range[2]) then
      return false
    end
  end
  return true
end

local function getCharEncoding(bytes, from, bytes_size)
  for _,encoding in ipairs(ENCODINGS) do
    local to = from + encoding.size - 1
    if to <= bytes_size then
      for _,row in ipairs(encoding.ranges) do
        if matchRanges(row, bytes, from, to) then
          return encoding.name, encoding.size
        end
      end
    end
  end
  return 'unknown', 1
end

function utf_validator.getEncoding(str)
  local bytes_size = #str
  local bytes = {str:byte(1, bytes_size)}
  local encoding
  local from = 1
  repeat
    local new_encoding, size = getCharEncoding(bytes, from, bytes_size)
    encoding = encoding or new_encoding
    if new_encoding == 'ascii' and encoding == 'utf-8' or
       new_encoding == 'utf-8' and encoding == 'ascii' then
       encoding = 'utf-8'
    elseif new_encoding == 'unknown' or encoding ~= new_encoding then
      return 'unknown'
    end

    from = from + size
  until from > bytes_size

  return encoding
end

return utf_validator
